### PR TITLE
OSAC-3771: implement UnassignHost for BCM inventory backend

### PR DESCRIPTION
**Pull request by mennyaboush**
## Summary
- Implement `BCMClient.UnassignHost` to release a BCM-managed host back to the pool and delete the on-demand BMH CR during BareMetalInstance deletion
- Extend `BCMAPI` interface with `GetDevice` and `UpdateDevice`; add `BMHLifecycleManager` interface for testability (same abstraction pattern as `BCMAPI`)
- BCM update (clear `osac_instance_id` + labels) before BMH deletion for crash recovery safety — both steps individually idempotent

## Design Reference
- [OSAC-1339 design doc](../enhancement-proposals/enhancements/OSAC-1339-bcm-backend/design.md), lines 531–556
- Ordering (BCM first, BMH second) is deliberate per design doc failure handling section

## Deferred Items (tracked in IMPLEMENTATION-ROADMAP.md)
- **K8s events** (`BCMHostReleased`/`BCMHostAssigned`/`BCMConnectionError`): BMF operator has no EventRecorder today — will add in [OSAC-3767](https://redhat.atlassian.net/browse/OSAC-3767) (AssignHost) for all three events at once
- **ConsumerRef ownership check**: crash-recovery edge case impossible until AssignHost exists — deferred to [OSAC-3767](https://redhat.atlassian.net/browse/OSAC-3767)

## Jira
[OSAC-3771](https://redhat.atlassian.net/browse/OSAC-3771)

## Test plan
- [x] 13 unit tests covering: normal unassign, selective label removal, empty labels, device not found, nil/empty extra_values, already unassigned, invalid host ID, empty host ID, GetDevice error, UpdateDevice error, UpdateDevice validation failure, DeleteBMH error
- [x] `make fmt` — no drift
- [x] `make lint` — 0 issues
- [x] `make build` — passes
- [x] `make test` — all pass, no regressions
- [x] `make manifests generate` — no drift

---

_This PR description was drafted with AI assistance ([create-pr](https://github.com/osac-project/osac-workspace/tree/main/skills/create-pr) v0.6.0). Review for accuracy_

## Summary by CodeRabbit

* **New Features**
  * Added support for unassigning hosts from inventory devices.
  * Host unassignment removes associated instance and supplied label metadata.
  * Corresponding bare-metal host resources are deleted after successful unassignment.
  * Repeated or already-unassigned requests are handled safely.

* **Bug Fixes**
  * Improved handling and reporting of missing devices, invalid host IDs, and inventory or resource errors.

### CODE REVIEW

**Comment by eranco74** [comment]
@coderabbitai full-review

File: bare-metal-fulfillment-operator/internal/inventory/bcm.go [hunk 1]
```diff
@@ -218,9 +226,53 @@ func (c *BCMClient) AssignHost(_ context.Context, _ string, _ string, _ map[stri
 	return nil, fmt.Errorf("bcm AssignHost not implemented")
 }
 
-// UnassignHost is implemented in OSAC-3771.
-func (c *BCMClient) UnassignHost(_ context.Context, _ string, _ []string) error {
-	return fmt.Errorf("bcm UnassignHost not implemented")
+// UnassignHost clears the OSAC assignment from a BCM device and deletes the
+// on-demand BMH CR. Ordering is BCM update before BMH deletion for crash
```
Comment (carbonin) [inline bare-metal-fulfillment-operator/internal/inventory/bcm.go:344]: >deletes the on-demand BMH CR

Should this be `UnassignHost`'s responsiblity? I expected this to be just the inverse of `AssignHost`.
Where does the BMH get created?

Comment (mennyaboush) [reply bare-metal-fulfillment-operator/internal/inventory/bcm.go:344]: The design required BMH create and delete on demand on the inventory side... the creation of the BMH at the assign is WIP

File: bare-metal-fulfillment-operator/internal/inventory/bcm.go [hunk 2]
```diff
@@ -218,9 +226,53 @@ func (c *BCMClient) AssignHost(_ context.Context, _ string, _ string, _ map[stri
...
+
+			if _, err := c.client.UpdateDevice(ctx, raw); err != nil {
+				return fmt.Errorf("UnassignHost: %w", err)
+			}
+		} else {
+			log.Info("BCM host already unassigned, skipping extra_values cleanup", "hostname", hostname)
+		}
+	} else {
+		log.Info("BCM device not found, skipping extra_values cleanup", "hostname", hostname)
+	}
```
Comment (carbonin) [inline bare-metal-fulfillment-operator/internal/inventory/bcm.go]: I don't like the way this has so many else blocks and then drops through to the DeleteBMH call.

I think this would be better as a separate function, that way the noop cases could be guards with an early return rather than these else blocks.

File: bare-metal-fulfillment-operator/internal/inventory/bcm_test.go [hunk 3]
```diff
@@ -40,6 +40,32 @@ type mockBCMAPI struct{}
 
 func (m *mockBCMAPI) CertWatcher() *certwatcher.CertWatcher                    { return nil }
 func (m *mockBCMAPI) GetDevices(_ context.Context) ([]bcmclient.Device, error) { return nil, nil }
+func (m *mockBCMAPI) GetDevice(_ context.Context, _ string) (*bcmclient.Device, error) {
```
Comment (carbonin) [inline bare-metal-fulfillment-operator/internal/inventory/bcm_test.go]: This will conflict with the change to mockgen in the other PR, right?

Comment (mennyaboush) [reply bare-metal-fulfillment-operator/internal/inventory/bcm_test.go]: Right. will move to mockgen here also

**Comment by mennyaboush** [comment]
Re: **nil-deref** bot finding — In the BCM backend, bmhManager is always non-nil. Metal3 management is required alongside BCM inventory. The Helm chart (PR #391) makes bmhNamespace required when BCM is enabled, and createBCMInventoryClient only runs when inventory type is bcm, at which point the management config is always metal3. A nil bmhManager would mean a misconfigured deployment that fails earlier in the startup path. Not adding a nil guard to keep the code clean.

**Comment by mennyaboush** [comment]
Addressed all review comments:

1. **"Where does BMH get created?"** — See PR #477 ([OSAC-3768](https://redhat.atlassian.net/browse/OSAC-3768)), which implements `createBMHAndBuildHost` in `AssignHost`. `UnassignHost` deleting the BMH is the inverse of `AssignHost` creating it.

2. **"Too many else blocks"** — Extracted `clearBCMAssignment` with early-return guards, no else blocks. `UnassignHost` calls it then `DeleteBMH` as two clean sequential steps.

3. **"This will conflict with mockgen"** — Rebased onto main (post-#463 merge), all tests use mockgen. No hand-written mocks.

**Review by carbonin — APPROVED** [review]
